### ISSUE TITLE
feat: Support eth address two-way attestation

### ISSUE THREAD

**Issue by NunoAlexandre** [radicle-dev/radicle-upstream#965]
**Depends on:** https://github.com/orgs/radicle-dev/projects/17 the integration of radicle-link's `next` branch.

This feature has an impact the whole Funding experience we are building by linking user's Radicle identities and their Ethereum accounts, leveraging discoverability (users will be able to tag each other in the Funding features by name instead of by addresses) and trust (users can trust that an address is associated with a given Radicle user and vice-versa).

### What needs to be done

- [ ] Add a "ethAddress" field to the Radicle Identity document (metadata?) or have a dedicated document for this purpose
- [ ]  Create the necessary infrastructure in `proxy` to set and get this field (http endpoint, etc)
- [ ] Wire up the attention flow in the frontend (see below)

#### Attestation flow

1. As a Radicle user, I have no registered eth address
2. I want to access the Radicle Funding features, therefore I go to "Funding"
3. In "Funding", I see a prompt to connect my wallet using the Eth account I want to use to register in the Radicle network with
4. I connect my wallet app using the QR code provided by Radicle
5. Now on Radicle, I see the correct account address displayed
6. I confirm I want to associate that eth account with my Radicle identity (attestation step 1/2)
6.2 I confirm a transaction storing that association on-chain (attestation step 2/2), which will prompt my wallet app for submission
7. I confirm the registration in my wallet app
9. the app confirms the registration is successful. Now I have access to all funding features.

10. Next time I connect using another account, I get a warning informing me that I am using an account that's not the one I have registered

You can see the stubbed flow from step 5 until 9:

![Peek 2020-12-03 16-59](https://user-images.githubusercontent.com/7075260/101055336-180a0180-358a-11eb-8ffc-5876ce41b208.gif)

**Comment by juliendonck** [comment]
@NunoAlexandre is this still relevant with the latest thinking?

**Comment by NunoAlexandre** [comment]
The issue is still a relevant one but out of scope for the Funding Pool MVP.

**Comment by xla** [comment]
@NunoAlexandre Would it be possible to capture this work somewhere else or do you think it has value to keep the issue around?

**Comment by NunoAlexandre** [comment]
@xla @juliendonck I have updated the description and scope of this issue to fully capture what it entails.



> @NunoAlexandre Would it be possible to capture this work somewhere else or do you think it has value to keep the issue around?

This issue is part of the `Funding` project board and the B2 milestone. Both seem good with me. I see value in leaving the issue around, yeah.

**Comment by CodeSandwich** [comment]
We had a talk with Kim and Nuno and we came up with some conclusions.

@cloudhead , could you take a look? You were working on that before and you've designed something slightly different, you may find some holes.

@kim , could you comment on `Storing identity document revision on chain` and `Many-to-many`? These are my thoughts from after the talk and I want to verify that.

@nuno, could you comment on `Many-to-many`? It shifts how we view the attestations.

# Solution

### Ethereum

Deploy a smart contract, which has two functions:

```
function attest(bytes32 linkId) public {
    emit Attested(msg.sender, linkId);
}

function revoke(bytes32 linkId) public {
    emit Revoked(msg.sender, linkId);
}
```

### Link

Store in the identity document a list of attested ethereum addresses. They can be added or removed with no restrictions.

# Rationale

## Ethereum events are enough

The attestations don't need to be accessed from smart contracts, so events are enough while being ~100 times cheaper. Neither events nor storage protect us from malicious Ethereum clients.

## Cross-storage of IDs is enough

Existence of an event is a proof that the holder of the ethereum private key accepted connection to the radicle ID. Existence of an ethereum address in the radicle user's identity document is a proof that the holder of the private key accepted connection to the ethereum address. If both sides are connected to each other at the given moment, attestation is obviously valid.

## Storing signatures is pointless

No matter how sophisticated data we sign with the private key from network A and store on network B, the users of network B won't be able to say anything sure about the state of network A without directly querying it. If I make an attestation on link and store the proof of that action on Ethereum, you won't be able to tell if the attestation on link is still valid just by looking at that proof. You'll need to query link and at this point the complex signature is useless.

## Split brain

The system should accept one-sided attestations, but not respect them. The period when attestation is in progress, the network peers see different states, attestation is single sided revoked due to lost keys, and simple mistakes will be occurring and should be tolerated.

## Storing identity document revision on chain

I don't think that we need the identity document revision on chain. To verify an ethereum address we should be looking ONLY at the freshest version of this document. The past state is irrelevant, if one looses an Ethereum private key, an update on link must be enough to revoke the attestation. It's already designed like this on Ethereum side, an attestation is valid only until it's revoked.

## Many-to-many

### Ethereum

I think that we should allow attesting on a single ethereum address multiple link IDs. A single treasury should be able to benefit from multiple link users.

### Link

I think that we should allow attesting on a single link ID multiple ethereum addresses. It kind of makes sense to include my address and the address of my favourite charity. It could also be used during a transition period when migrating from one address to another. There may be other use cases, e.g. something beyond simple donations. Pragmatically I think that we should make it possible to store multiple Ethereum addresses in the link identity document, but for now only use the first valid attestation in the upstream as having multiple would complicate things.

**Comment by NunoAlexandre** [comment]
Thanks for the write-up, @CodeSandwich :raised_hands: 

So regarding the many-to-many, we have to keep in mind the application of this feature. Specifically, discoverability.

Say that user A wants to support user B. 

1. If all user A has is the Radicle ID of user B:
  a. that doesn't seem any better, from a UX standpoint, than only having their eth address.
  b. if that Radicle ID can have many eth addresses associated, how would we resolve which one to pick?
      maybe the answer is "it wouldn't matter", but from a UX perspective it's something we need to consider.
      we could simply pick one deterministically and hide the many-to-many factor from the UI, but it does
      introduce complexity. Just something to keep in mind.
      
2. If all user A has is the Eth address of a user B:
   a. when user A would input that address to add it as one of their receivers, the idea is to have us looking up
       the respective associated Radicle user. Now, if that address might be associated with many Radicle Ids,
       that would be equally confusing. In the end, the address is the final destination of the support, so if user A
       knows that address to be from the user they know/expect that'd be all good, but it literally could happen that
       user A inputs `0x8a12...0F12` and they'd get `Nuno, Linus Torvalds, Simon Peyton Jones`.

When yesterday I mentioned the ENS component was precisely to avoid these scenarios. If we could define a global
name that would disambiguate. The user would be able to associate a name to (many) their radicle identities(s) and Eth address(es).

> It kind of makes sense to include my address and the address of my favourite charity.

I would not intertwine these two things. Having users 'registering' their addresses is a separate concern from proxying donations.
For donations, we have the Funding pool and we have the idea of the 'awesome lists', where your _proxy_ feature in the contracts comes into play.

**Comment by CodeSandwich** [comment]
That's a good point and you're right, it would complicate the whole system to an extreme degree with multiple edge cases without providing much value. I'll drop it.

I was thinking about resilience of the network. With 1-1 attestation it's very easy for link to diverge from ethereum and break the attestation for some people. But it's exactly the same with Twitter, GitHub and all the other identity attestations. If we ever want to and are able to solve the divergence problem for these services, it'll automatically fix for ethereum, no need for special treatment.

**Comment by kim** [comment]
Okay... I'm not sure if we understand / agree on what the problem is we're
trying to solve here, and what the tradeoffs are. So let's take a step back
again, and try to avoid jumping to conclusions in a dismissive way:

We are talking about individuals: someone has a radicle link identity of kind
person, and wants to attach an address to it, so they can receive money. We are
**not** talking about projects, which can have multiple stakeholders.

Alright, so why do we need any kind of cryptography here? After all, if I put a
donate-via-paypal button on my homepage, nobody is going to ask paypal for a
proof of whether that's actually my homepage, right?

Why not? Because we can make this reasonably secure using the internet's
predominant trust model: I could purchase an EV certificate, convince the
certificate provider that I own a particular email address, and announce that
I can receive paypal payments through that. Anyone can look at the whois
database to check that the email address matches.

Why would we do that? Because someone could break into my webserver, modify my
htmls, and put their own email address there.

Okay, yeah, that's a bit contrived for practical purposes, but you get the idea:
because there is no webserver, anyone can modify the htmls along the way until
they end up on your machine. By signing all sorts of things, we can defend
against such tampering, so one can be reasonably sure that what they received is
indeed what I authored.

Yet... how do you know its me? Answer: you have to TOFU my public key(s), or
received them out-of-band through a secure channel. This is kind of crucial,
because literally everyone on the radicle-link network could pick "nunito" as
their display name. Or "malkovich".

So what was it again we wanted to defend against? Ah right: in the absence of a
trusted authority **as well as** an out-of-band key exchange, Teh Giver of Rad
might mistakenly pick a scammer's address, just because the display name looks
the same. Unlike in the fiat world, there is no way to reverse a transaction, so
that's kind of bad.

Alright, so the problem boils down to: either I trust a particular Ethereum
address (or an ENS name), or I trust a particular link ID, and from there can
extend trust to the other side by simply a claim on the trusted side. ENS is
kind of attractive here, because it needs to be renewed, which avoids some weird
states due to abandoned accounts (within the expiration window, of course). The
same could be employed in link. Note that there is no additional value in a
cross-reference in this scenario, because we have already established trust (see
the paypal example).

The problem becomes more interesting if we assume that we **kind of** trust one
side, but want to **increase** our confidence by obtaining a proof of key
ownership from the respective other side, which, together with the proof on this
side, constitutes non-repudiation (which is what we're after). It is a
non-solution to just look at a non-authenticated claim value on both sides and
check that they are equal because a. there is no correlation of causal history
between both chains, and b. it implies that full verification is performed on
both sides. b. is particularly bad in our setting -- if we'd force users to
launch an Ethereum light client alongside a link node every time they start
`upstream`, they'd be like uh yeah an electron bloat mining bitcoins thanks bye
report-as-violates-appstore-tos.

There is no simple solution to this, because all we can do is to **gradually**
increase trust through multiple means. We can, for example, look at the tracking
graph in link and infer a web-of-trust from that. We can allow Keybase-style
proofs to be stored in link. And, and that's what we're discussing, we can
publish Keybase-style proofs on external systems, including Ethereum. If we
assume that we are not actually verifying block headers, we can still obtain a
cryptographically verifiable proof that the key owner had access to _some_
transaction key which produced either a state transition or some outputs.

This

> ## Ethereum events are enough

is thus the only interesting thing in what Igor wrote, even though he missed out
on the opportunity to answer what they are enough for. Iiuc, "events" is the
Solidity name for logs, which are part of transaction receipts. Therefore, it
should be possible to compute a merkle proof of whether a particular "event" was
included in a block. If the receipt and the block header come from two different
sources, that seems like a great way to have a lightweight plausibility check
without requiring full header verification.


What is still unclear to me however, even after many rounds on this topic, is
whether the feature at hand requires anything more than a claim on-chain that a
particular radicle ID is associated with an account. We have this kind of DApp
frontend for our set of contracts, and if I'm crypto-savvy enough to consider
actually transferring money, I will (hopefully) know how to do my due diligence.
There isn't a need to publish the address on the link side, because we can make
the association by only looking at the chain.

I suppose that's the actionable item for @cloudhead and @lftherios to answer.

**Comment by CodeSandwich** [comment]
Thank you for the detailed response. IIUC it boils down to the fact that we want 2 levels of trust for attestation:

- Level 1: I have access only to network A:
  - I can prove that the owner of identity X on network A attests identity Y from network B at the current moment modulo freshness of my network A data.
  - I can prove that the owner of identity Y on network B attested identity X from network A at some point in the past and I know when it has happened.
- Level 2: I have access to both networks:
  - I can prove that owners of both identities attest each other at the current moment modulo freshness of my networks' data.

Do we need to be able to extract or prove more information?

The designed I've proposed provides only the second level. I agree that if we want to support the first level as well, we'll need some signatures magic. Let me think for a moment, how could that work.

**Comment by CodeSandwich** [comment]
Thank you for the enormous explanation, it's very helpful.

Indeed we must account for lack of trust for any network outlet on either side. Luckily on link we can prove that an identity document was updated by the holder of a given ID and on ethereum we can prove that a transaction was submitted by an owner of a given address. A malicious outlet can try to hide an action, but it can never fake it.

> to just look at a non-authenticated claim value on both sides and check that they are equal because (...) it implies that full verification is performed on both sides

That's not necessary, if we have a proof that A attests B and another one that B attests A, it's obvious that they attest each other. We don't need a custom ethereum client for that, if the signed transaction was handed to me by Al Capone, I would still trust its origin.

> there is no correlation of causal history between both chains

How is that a problem? We don't care about the history, only the latest state is important. The only problem here is that the attestation can be revoked or changed and a malicious outlet can try to hide or delay that event, but that's an unavoidable problem in a distributed system.

---

I think that there's another trick worth pursuing: **single sided proof of attestation**. It would be nice to be able to somehow verify attestation without querying the other network. That forces us to store a signature from the other side as a proof that the holder of that key approves what we're doing.

As an extra feature it could be beneficial to make both sides accept and sign a timestamp of the start and end of an attestation, so it could have perfectly synchronized transitions and die if abandoned. Unfortunately `<label>.radicle.eth` ENS domains don't need renewal. The rationale is that an expired yet widely trusted domain would pose a danger.

**Comment by kim** [comment]
> the signed transaction was handed to me by Al Capone, I would still trust its origin

In your proposal the transaction stores (or outputs) a random value, there is nothing you can say about it unless you either trust the transaction key already (if that's what you mean by "origin"), or you trust the radicle ID (if that's what you mean by "origin"). Otherwise, Al Capone can just sign with a random key, and you would trust it.

If you already established trust, however, then looking for a(nother) random value on the other side doesn't add much.

Consider the following: say we model this as a name service. Someone thinks they're smart and squats "linus", claiming that `rad:git:hyydeathbyfa1afe1` is the corresponding radicle ID. If you have convinced yourself that "linus" is indeed the person you think he is, a reverse confirmation is not needed. If you are unsure about that, you need to go and convince yourself that `rad:git:hyydeathbyfa1afe1` is owned by the person you're expecting. So how do you do that? You can download the identity history, verify it, and look for hints that this profile is somehow legit. But what if the squatter actually owns the keys? They can craft the profile in a way that makes it _look_ legit, perhaps including the eth address. This means we haven't won anything _unless_ we have a way to obtain the public key(s) from the real Linus through some other channel, or have other evidence (such as keybase-style proofs) which convinces us that it is _most likely_ Linus.

But if we trust `rad:git:hyydeathbyfa1afe1`, why do we need to confirm that through "linus"?


You see, we have to start with a trust root somewhere. Given the controversy about crypto in general, I would expect that early users of our funding feature are already crypto-savvy, and will find ways to make their ownership of a particular ethereum address plausible to their peers. So we could just start there.

When it comes to crypto sceptics, radicle-link would greatly benefit from keybase-style proofs, because, as I wrote, this helps increasing trust _without_ having to trust the service which relays the proof. For this to work, you must funnel a signature through that service, which you can verify with what you already have. Note that the proof remains valid even if the set of keys changes, because we have a "sigchain". Also note that revocation happens on one side only -- if the service denies relaying the proof, then well, you have one less credit in your trustworthiness meter, but nothing more. Lastly note that a blockchain is no different from Twitter or Reddit or whatever -- its role is to relay a proof which we can check with the information we already have, and nothing else.


> How is that a problem? We don't care about the history, only the latest state is important.

"Latest state" is another way of describing consensus in distributed systems. If you send a transaction to a distributed ledger, it will be finalised on top of the latest block of the chain, or not at all. If you create a commit in a git repository, you can pick any parent. If you own a quorum of the keys of a radicle link identity, you can create a "fork" of your identity which is equally valid.

Therefore, if you want to create a way to say "at this point in time, something something", you need to say what time it is. That is, you need to tell the timestamping service (the blockchain) what time it is in git, by including the hash value which constitutes it.

> As an extra feature it could be beneficial to make both sides accept and sign a timestamp of the start and end of an attestation, so it could have perfectly synchronized transitions and die if abandoned.

I don't think "perfectly" is the right word here, but yes, I think it would be wise to expire claims unless renewed. So this

> Unfortunately <label>.radicle.eth ENS domains don't need renewal. The rationale is that an expired yet widely trusted domain would pose a danger.

is unfortunate indeed. Can't those domains have a large period of time (say, a year) in which they're locked if they weren't renewed?

**Comment by CodeSandwich** [comment]
> Al Capone can just sign with a random key, and you would trust it.

He can't. If your identity link states that your ethereum address is `0x1234...`, I can ask Al about events coming from attestations of `0x1234...`. I did some research and now I'm almost certain that the events aren't proofs of anything unless you trust the client. But in our case it's not a problem, I can look up the transaction, which supposedly generated the event and it will contain the address of the attestation contract, the inputs and the signature proving that the whole transaction has been created by the holder of the private key controlling address `0x1234...`.

Overall at this point I agree that we need to store on network A things signed with keys from network B for their self-containedness. But I want to understand how things are working and why things must be designed the way they are.

> we have to start with a trust root somewhere.

That's a big overarching problem, but for now let's focus on fixing this little part of the solution with a two-way attestation.

> you can create a "fork" of your identity which is equally valid

Aah, that's where storing the revision on ethereum comes in handy! Only people respecting that revision and its children should acknowledge the attestation. IIUC only one, newest attestation on ethereum for one revision of a link identity should be considered?

**Comment by kim** [comment]
> I did some research and now I'm almost certain that the events aren't proofs
> of anything unless you trust the client. But in our case it's not a problem, I
> can look up the transaction

I don't know Igor... what does "look up the transaction" mean? What you got from
Al is a transaction receipt. It contains the block hash, so you can go get the
block header from somewhere else, figure out if the receipt is included in the
merkle root -- and then you haven't trusted anyone, but you need to do more
work.


> IIUC only one, newest attestation on ethereum for one revision of a link
> identity should be considered?

Mmya... isn't the idea that there can only be a 1-1 mapping between ethereum
address and link id? So the root hash (which is the id) must be unique. If we
allow to renew the attestation with a newer revision, we can't verify on-chain
that the previous one is indeed an ancestor.

**Comment by CodeSandwich** [comment]
I imagine that the interactions with a 3rd party client would be this, the JSON RPC API is standardized:
- Use https://eth.wiki/json-rpc/API#eth_getfilterlogs to get a list of interesting events and corresponding `transactionHash`es
- Fetch the transaction data by `transactionHash` with https://eth.wiki/json-rpc/API#eth_gettransactionbyhash
- Validate that the signature inside matches the data and the ethereum address. For reference the signature payload content is listed here: https://eth.wiki/json-rpc/API#eth_signtransaction, all these fields are present in the transaction data.
- Now we know that this specific set of parameters has been signed by the owner of the ethereum address, we can use it

**Comment by kim** [comment]
https://github.com/radicle-dev/radicle-link/pull/525

**Comment by NunoAlexandre** [comment]
Quoting the latest update on this work:


<blockquote>
It has been concluded that for the use-case of radicle-dev/radicle-upstream#965, a simple one-way attestation in the other direction suffices, and can be expressed with a much more compact proof:

Consider a contract which allows to claim (and unclaim) some Radicle ID (URN). No two claims for the same ID at the same time are permitted. After a successful claim, the _transaction ID_ is included in the claimed URN's identity history.

If either the Ethereum address _or_ the `radicle-link` identity is trusted, only the inclusion of the transaction in a block on the main chain needs to be verified. Navigation from address to URN entails inspecting the transaction payload. Navigating from `radicle-link` identity to address entails inspecting the transaction sender.

Note that no elevation of trust results from this, nor assurance of an association with a "real" person. Neither does it improve the fork resiliency of the `radicle-link` identity, nor defend against hiding of updates. I therefore think that this RFC has a justification in its own right.

Since the identity document payload can be freely extended by library users, and both the payload and the verification semantics are entirely defined by the external system, I don't think a subsequent RFC is required. I also don't think any changes to `radicle-link` are required, except perhaps for convenience or code organisation reasons. #464 would not hurt, though.

by @kim (https://github.com/radicle-dev/radicle-link/pull/525#issuecomment-789185194)
</blockquote>

**Comment by NunoAlexandre** [comment]
With https://github.com/radicle-dev/radicle-upstream/issues/965#issuecomment-790616025 in mind, and follow up discussions on what we need to build to suffice our needs, the flow will look like this:

![flow](https://user-images.githubusercontent.com/7075260/109980630-643b1e00-7d00-11eb-852b-eea2e6f90702.jpeg)


1. The **user** stores, on-chain, a _pointer_ to their **radicle identity** [`register`]
   ⓘ By _pointer_, we mean a minimal amount of data we can store and be able to find the associate radicle identity. So it could be the radicle id itself, but that would be too much data. We spoke of storing the root of the identity instead.
 
   → said user pays a **transaction** fee to register this claim
   ← as a return value of this operation, we hold on to the id of said transaction
    
2. The **user** stores the transaction id of the previous step in their radicle identity
   ⓘ By doing so, when all we have is a Radicle identity, we have a transaction id that
       allows us to look up a transaction and check its claim.

   → said transaction id is added to the Radicle identity 
   ← as a return value of this operation, we have a new revision of said identity

### :heavy_check_mark:  Registered

With these two steps complete, we have a distributed state where a Radicle user claims, on-chain, to be (the owner of a) Radicle identity, which is thus associated with the eth address involved. On Radicle Link, we have a transaction Id that we can use to look up and check whether said identity matches the claim that is found on-chain. If they match, the attestation is considered valid. if not, it is naturally considered invalid.

### :key: Lost keys
 
Now, what happens if we lose either or both of the keys, i.e, the Radicle Identity keys and/or the Eth account private key?

1. The user loses their Radicle identity (keys)
   
    In this case, the Ethereum account that registered the claim in **step 1** of the attestation is assumed to be still in the user's possession. In that case, the user _could_ **unregister** their claim on-chain.

    However, if the user intends to completely abandon the platform, there is no incentive for them to pay yet another transaction fee. Something for us to keep in mind.

2. The user loses their Ethereum account key

   In this case, the user **U** will **not** be able to **unregister** their claim on-chain. This orphan state is undesired as it leaves the user impotent and grants no way for, within the Radicle network, for revocation, which would lead to, say, having _n_ users supporting **U** without knowing that their support is falling into a vacuum.

   To reduce the strain of such a scenario, we include an **expiration date** to the claim stored in the Radicle identity. More on that below.

3. The user loses both their Radicle identity and their Eth account keys
  
   In such a case, similarly to the previous scenario, said **expiration date**  will, eventually and automatically, revoke the attestation.


### Takeaways :takeout_box: 


1. On-chain, we only store a _pointer_ to the user's Radicle Identity

2. On the Radicle identity, we store the tuple `(transaction-id, expiry-date)`
   
   Given the `expiry-date` part of the claim in the Radicle identity, 
   the user _must_ renew the claim within that period to avoid the
   attestation from becoming invalid. This means that, in the app,
   we need to keep an eye on this value and prompt the user to renew
   the claim, resulting in a new revision of the identity being created with
   a refreshed `expiry-date`.

   This is something that we need to work on with @radicle-dev/product :point_up: 

3. Therefore, an attestation is considered valid _iff_:
   
     ```
       all 
         [ User **U* has a claim, on-chain, that holds a pointer to their Radicle Identity
         , User **U** has a `(transaction-id, expiry-date)` in their Radicle Identity,
         , `expiry-date >= now()`
         , the transaction indexed by transaction-id points to said Radicle identity
         ]
   ```


@kim @CodeSandwich please let me know if I missed anything
@cloudhead your beautiful eyes would be much appreciated here

**Comment by kim** [comment]
Thanks @NunoAlexandre 

Two additions:

* nit: the radicle ID _is_ the root hash, but we can omit type tags (git, hash algorithm), if we can introduce them later. The hash algorithm is more crucial than the protocol identifier, because it would change the size of the data that needs to be stored.
* Unregistering is possible without locking the radicle id for eternity. Thus, in addition to the above, the verification must check that the _current_ address (as in, latest known chain state) matches the address of the transaction whose hash is in the radicle-link document.

**Comment by CodeSandwich** [comment]
Excellent writeup, @NunoAlexandre!

I'll add one more nit:
- We probably should store the ethereum address in the claim on link. As we've discussed, a historical transaction doesn't proof anything, it can be long revoked by now, we need to query for events and check what does the newest one say.

**Comment by NunoAlexandre** [comment]
> Excellent writeup, @NunoAlexandre!
> 
> I'll add one more nit:
> 
> * We probably should store the ethereum address in the claim on link. As we've discussed, a historical transaction doesn't prove anything, it can be long revoked by now, we need to query for events and check what does the newest one say.

Thanks for the feedback, @CodeSandwich!

To make sure I am understanding your point well, you are saying that just looking up the transaction **T** (of which we hold the id) is not meaningful since a newer transaction could have come into play that could unregister the claim over the radicle id.

That point seems connected to Kim's second point above. Indeed, it's a good point, we do need to query the chain for potential newer transactions issued by that same address. But I wouldn't know if we _need_ to store the address in the radicle identity, since we could retrieve it when we look up **T**. It could be useful to store it, to avoid this roundtrip, but since we have to lookup **T** anyway, we could probably save the redundancy.

**Comment by CodeSandwich** [comment]
Yes, it's deeply connected to Kim's second point.

If I understand correctly, the current proposal requires doing these steps in order to check if a claim found on link is confirmed on Ethereum:

1. Look up transaction by hash **T**
2. Verify the signature and extract the sender **U**
3. Query for the latest event from confirmation/revocation made by **U**
4. Look up transaction which supposedly generated this event
5. Verify the signature and content of the transaction

If we store the Ethereum address **U** directly in link, we can skip steps 1 and 2.

I'm making an assumption, that a single Ethereum address can claim only one radicle ID, which means that the latest claim invalidates all previous ones. IIUC that's what we've agreed on previously in this thread.

**Comment by CodeSandwich** [comment]
Yes, it's deeply connected to Kim's second point.

I'm making an assumption, that a single Ethereum address can claim only one radicle ID, which means that the latest claim invalidates all previous ones. IIUC that's what we've agreed on previously in this thread.

If I understand correctly, the current proposal requires doing these steps in order to check if a claim found on link is confirmed on Ethereum:

1. Look up transaction by hash **T**
2. Verify the signature and extract the sender **U**
3. Query for the latest event from confirmation/revocation transaction **T2** made by **U**
4. Look up transaction **T2** which supposedly generated this event (skip if **T** == **T2**)
5. Verify the signature and check if the transaction claims the right radicle ID

If we store the Ethereum address **U** directly in link, we can skip steps 1 and 2. **T** may or may not be the source of the latest event. If it isn't, we end up fetching first **T** and then **T2**, which is wasteful, because we're going to respect only content of **T2**.

**Comment by kim** [comment]
> This means that, in the app,
we need to keep an eye on this value and prompt the user to renew
the claim, resulting in a new revision of the identity being created with
a refreshed expiry-date.

I’m wondering if this can be made safe even if one uses some other tool. As in, you cannot update an identity which contains an expired entry, and you cannot convert it into a “verified” type. Bit of a modularity layer cake conundrum, because that specific payload must be known. /cc @FintanH @xla

**Comment by rudolfs** [comment]
@CodeSandwich is this feature now fully implemented? Can we close this issue?

**Comment by CodeSandwich** [comment]
Yeah, I think so
